skyportal/handlers/api/gcn.py

```diff
@@ -32,6 +33,98 @@
 Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
 
 
+def post_gcnevent(payload, user_id, session):
+    """Post GcnEvent to database.
+    payload: str
+        VOEvent readable string
+    user_id : int
+        SkyPortal ID of User posting the GcnEvent
+    session: sqlalchemy.Session
+        Database session for this transaction
+    """
+
+    user = session.query(User).get(user_id)
+
+    schema = f'{os.path.dirname(__file__)}/../../utils/schema/VOEvent-v2.0.xsd'
+    voevent_schema = xmlschema.XMLSchema(schema)
+    if voevent_schema.is_valid(payload):
+        # check if is string
+        try:
+            payload = payload.encode('ascii')
+        except AttributeError:
+            pass
+        root = lxml.etree.fromstring(payload)
+    else:
+        raise ValueError("xml file is not valid VOEvent")
+
+    dateobs = get_dateobs(root)
+
+    try:
+        event = GcnEvent.query.filter_by(dateobs=dateobs).one()
+
+        if not event.is_accessible_by(user, mode="update"):
+            raise ValueError(
+                "Insufficient permissions: GCN event can only be updated by original poster"
+            )
+
+    except NoResultFound:
+        event = GcnEvent(dateobs=dateobs, sent_by_id=user.id)
+        session.add(event)
+
+    tags = [
+        GcnTag(
+            dateobs=event.dateobs,
+            text=text,
+            sent_by_id=user.id,
+        )
+        for text in get_tags(root)
+    ]
+
+    gcn_notice = GcnNotice(
+        content=payload,
+        ivorn=root.attrib['ivorn'],
+        notice_type=gcn.get_notice_type(root),
+        stream=urlparse(root.attrib['ivorn']).path.lstrip('/'),
+        date=root.find('./Who/Date').text,
+        dateobs=event.dateobs,
+        sent_by_id=user.id,
+    )
+
+    for tag in tags:
+        session.add(tag)
+    session.add(gcn_notice)
+
+    skymap = get_skymap(root, gcn_notice)
+    if skymap is None:
+        return event.id
+
+    skymap["dateobs"] = event.dateobs
+    skymap["sent_by_id"] = user.id
+
+    try:
+        localization = (
+            Localization.query_records_accessible_by(
+                user,
+            )
+            .filter_by(
+                dateobs=dateobs,
+                localization_name=skymap["localization_name"],
+            )
+            .one()
+        )
+    except NoResultFound:
+        localization = Localization(**skymap)
+        session.add(localization)
+        session.commit()
```

Comment:
The only place where you are committing is inside an except clause. That doesn't look right...

Reply:
This is just checking for a repeat of the localization.

Reply:
Yeah but you are adding and modifying some rows in this function. Are you sure those changes are getting committed at all?

Reply:
@guynir42 its behaved that way for a long time now. I tested it repeats the old behavior.

Reply:
Hrm, what needs to be clarified is if each of the following sometimes raises a NoResultsFound exception:

```python
         event = GcnEvent.query.filter_by(dateobs=dateobs).one()
```

and

```python
            Localization.query_records_accessible_by(
                user,
            )
            .filter_by(
                dateobs=dateobs,
                localization_name=skymap["localization_name"],
            )
            .one()
```

Reply:
Given that the demo data and the usual flow works, my intuition is this is all fine.